# Проект: Рынок видеоигр


### Цели и задачи проекта
 
**Цель проекта:** Привлечение новой аудитории в игру "Секреты темнолесья" с помощью публикации статьи о развитии индустрии игр в начале XXI века на IT-ресурсе. 
В статье будет представлен обзор об играх - основной акцент на жанр RPG, а также проведен анализ игровых платформ, продаж игр разных жанров и региональные предпочтения игроков.

**Задачи проекта:** Изучить развитие игровой индустрии с 2000 по 2013 год. Для этого понадобится познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.

### Описание данных



Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- *Name* — название игры.
- *Platform* — название платформы.
- *Year of Release* — год выпуска игры.
- *Genre* — жанр игры.
- *NA sales* — продажи в Северной Америке (в миллионах проданных копий).
- *EU sales* — продажи в Европе (в миллионах проданных копий).
- *JP sales* — продажи в Японии (в миллионах проданных копий).
- *Other sales* — продажи в других странах (в миллионах проданных копий).
- *Critic Score* — оценка критиков (от 0 до 100).
- *User Score* — оценка пользователей (от 0 до 10).
- *Rating* — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта



1. Загрузка и знакомство с данными
2. Проверка ошибок в данных и их предобработка
- Названия, или метки, столбцов датафрейма
- Типы данны
- Наличие пропусков в данных
- Явные и неявные дубликаты в данных
3. Фильтрация данных
4. Категоризация данных

---

## 1. Загрузка данных и знакомство с ними



In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('/datasets/new_games.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


В датафрейме 11 столбцов и 16956 строк. Типы данных представлены float64 (4 столбца) и object (7 столбцов). Из них: 

- Для столбцов 'Name', 'Platform', 'Genre' выбран тип данных **object**, что подходит для представленных данных.
- Столбец 'Year of Release' представлен типом данных **float64** - дробные значения не включает, поэтому можно переделать в тип  **int64**
- Для столбцов 'EU sales' и 'JP sales' используется некорректный тип данных - **object**. Так как данные столбцы указывваают кол-во проданных копий игр в миллионах лучше использовать тип данных **float64**. Столбцы 'NA sales' и 'Other sales' представлены корректным типом данных **float64**.
- Для столбца 'Critic Score' выбран верный тип данных **float64**. Данных тип даных также подойдет для столбца 'User Score', на данный момент он представлен типом данных **object**.
- Стобец 'Rating' хранит текстовые данные, но его можно рассмотреть как категориальный признак. В этом случае можно использовать тип **category**, чтобы улучшить производительность и оптимизировать память, если набор значений ограничен и заведомо известен.

Только в 5 столбцах из 11 нет пропущенных значений, а именно в 'Platform', 'NA sales', 'EU sales', 'JP sales', 'Other sales'. 


Все названия столбцов отражают содержимое данных. С целью поддержания правил хорошего стиля стоит переделдать названия столбцов в "snake case" (когда имя переменной состоит из нескольких слов, их пишут через знак подчёркивания),а также в именах переменных используют только строчные буквы.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма


In [5]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
df.columns = df.columns.str.lower().str.replace(' ','_')
print(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


### 2.2. Типы данных



In [7]:
#Удаляем пропуски и преобразуем 'Year of Release'
df = df.dropna(subset=['year_of_release'])
df['year_of_release'] = df['year_of_release'].astype('int64')

In [8]:
#Преобразуем 'EU sales' и 'JP sales'
df['eu_sales']=pd.to_numeric(df['eu_sales'], errors='coerce')
df['jp_sales']=pd.to_numeric(df['jp_sales'], errors='coerce')

In [9]:
#Преобразуем 'User Score'
df['user_score']=pd.to_numeric(df['user_score'], errors='coerce')


In [10]:
#Преобразуем 'Rating'
df['rating'] = df['rating'].astype('category')

In [11]:
#Выведем новый тип данных
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   name             16679 non-null  object  
 1   platform         16681 non-null  object  
 2   year_of_release  16681 non-null  int64   
 3   genre            16679 non-null  object  
 4   na_sales         16681 non-null  float64 
 5   eu_sales         16675 non-null  float64 
 6   jp_sales         16677 non-null  float64 
 7   other_sales      16681 non-null  float64 
 8   critic_score     8085 non-null   float64 
 9   user_score       7558 non-null   float64 
 10  rating           9901 non-null   category
dtypes: category(1), float64(6), int64(1), object(3)
memory usage: 1.4+ MB


### 2.3. Наличие пропусков в данных




In [12]:
#Кол-во пропущенных строк в датафрейме
df.isna().sum()

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8596
user_score         9123
rating             6780
dtype: int64

In [13]:
#Процент пропущенных строк
df.isna().sum() / len(df) * 100

name                0.011990
platform            0.000000
year_of_release     0.000000
genre               0.011990
na_sales            0.000000
eu_sales            0.035969
jp_sales            0.023979
other_sales         0.000000
critic_score       51.531683
user_score         54.690966
rating             40.645045
dtype: float64

In [14]:
#Доля пропущенных значений среди всех данных
df.isna().mean()

name               0.000120
platform           0.000000
year_of_release    0.000000
genre              0.000120
na_sales           0.000000
eu_sales           0.000360
jp_sales           0.000240
other_sales        0.000000
critic_score       0.515317
user_score         0.546910
rating             0.406450
dtype: float64

**Пропущенные значения**
- Данные пропущены в 7 столбцах ('Name', 'Genre', 'EU sales', 'JP sales', 'Critic Score', 'User Score', 'Rating')
- Максимальное значение пропущенных данных в столбце 'User Score' - 54%, на втором месте столбец 'Critic Score' - 51%, на третем месте столбец 'Rating' - 40%

- Пропуски в колонке 'Name' не критичны, поэтому их можно игнорировать или удалить, т.к их кол-во менее 1%.
- Пропуски в колонке 'Genre' не критичны, поэтому их можно игнорировать или удалить, т.к их кол-во менее 1%.
- Пропуски в колонке 'EU sales' и 'JP sales' можно заменить на среднее (в зависимости от названия платформы и года выхода игры)
- Пропуски в колонке 'Critic Score' не критичны, можно игнорировать.
- Пропуски в колонке 'User Score' не критичны, можно игнорировать.
- Пропуски в колонке 'Rating' не критичны, можно игнорировать.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [15]:
#Убираем пропуски в 'EU sales' - среднее значение в зависимости от названия платформы и года выхода игры
def group_eu_sales(row):
    group = df[(df['platform'] == row['platform']) & 
               (df['year_of_release'] == row['year_of_release'])]
    return group['eu_sales'].mean()
df['eu_sales'].fillna('group_eu_sales', inplace=True)

In [16]:
#Убираем пропуски в 'JP sales' - среднее значение в зависимости от названия платформы и года выхода игры
def group_jp_sales(row):
    group = df[(df['platform'] == row['platform']) & 
               (df['year_of_release'] == row['year_of_release'])]
    return group['jp_sales'].mean()
df['jp_sales'].fillna('group_jp_sales', inplace=True)

In [17]:
#Удаляем пропуски в 'Name' и 'Genre'
df = df.dropna(subset=['name', 'genre'])

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   name             16679 non-null  object  
 1   platform         16679 non-null  object  
 2   year_of_release  16679 non-null  int64   
 3   genre            16679 non-null  object  
 4   na_sales         16679 non-null  float64 
 5   eu_sales         16679 non-null  object  
 6   jp_sales         16679 non-null  object  
 7   other_sales      16679 non-null  float64 
 8   critic_score     8085 non-null   float64 
 9   user_score       7558 non-null   float64 
 10  rating           9901 non-null   category
dtypes: category(1), float64(4), int64(1), object(5)
memory usage: 1.4+ MB


### 2.4. Явные и неявные дубликаты в данных



In [19]:
unique_g = df['genre'].unique()
unique_p = df[ 'platform'].unique()
unique_r = df['rating'].unique()
unique_yor = df['year_of_release'].unique()
print(unique_g, unique_p, unique_r, unique_yor)

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY'] ['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX'] ['E', NaN, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP']
Categories (8, object): ['E', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'] [2006 1985 2008 2009 1996 1989 1984 2005 1999 2007 2010 2013 2004 1990
 1988 2002 2001 2011 1998 2015 2012 2014 1992 1997 1993 1994 1982 2016
 2003 1986 2000 1995 1991 1981 1987 1980 1983]


In [20]:
nunique_count = df[['genre', 'platform', 'rating', 'year_of_release']].nunique()
print(nunique_count)

genre              24
platform           31
rating              8
year_of_release    37
dtype: int64


In [21]:
df.columns = df.columns.str.lower().str.replace(' ','_')

#Приводим жанр к верхнему регистру
df['genre']=df['genre'].str.lower()

#Приводим рейтинг к верхнему регистру
df['rating']=df['rating'].str.upper()

In [22]:
#Наличие дубликатов
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
16951    False
16952    False
16953    False
16954    False
16955    False
Length: 16679, dtype: bool

In [23]:
# Сортируем датафрейм по всем столбцам
df_sorted = df.sort_values(by=df.columns.tolist())

# Находим дубликаты
duplicates = df_sorted[df_sorted.duplicated(keep=False)]

print(duplicates)

                                   name platform  year_of_release  \
15191                    Beyblade Burst      3DS             2016   
15192                    Beyblade Burst      3DS             2016   
15301                 11eyes: CrossOver     X360             2009   
15302                 11eyes: CrossOver     X360             2009   
4860   18 Wheeler: American Pro Trucker      PS2             2001   
...                                 ...      ...              ...   
2909   Yu-Gi-Oh! The Falsebound Kingdom       GC             2002   
6695                      Zoo Resort 3D      3DS             2011   
6696                      Zoo Resort 3D      3DS             2011   
8156                 Zumba Fitness Rush     X360             2012   
8157                 Zumba Fitness Rush     X360             2012   

              genre  na_sales eu_sales jp_sales  other_sales  critic_score  \
15191  role-playing      0.00      0.0     0.03         0.00           NaN   
15192  role-pla

In [24]:
#Кол-во дубликатов
df.duplicated().sum()

235

Всего в датафрейме 235 явных дубликатов. Их можно удалить.

In [25]:
df_no_duplicates = df_sorted.drop_duplicates()

In [26]:
#Кол-во строк до удаления дубликатов
df_row_count = df.shape[0]
print(df_row_count)

16679


In [27]:
#Кол-во строк после удаления дубликатов
final_row_count = df_no_duplicates.shape[0]
print(final_row_count)

16444


In [28]:
#Всего удалено строк (абсолютное кол-во)
print(df_row_count - final_row_count)

235


In [29]:
#Процент удаленных строк(относительное кол-во)
(df_row_count - final_row_count)/df_row_count*100

1.4089573715450565

Всего в датафрейме выявлен 235 явных дубликатов, что составило ~1.41% от всего датафрейма. После проведения нормализации данных с текстовыми значениями и удаления явных дубликаитов осталось 16444 строк. 

---

## 3. Фильтрация данных

- Данные за период с 2000 по 2013 год включительно. 

In [30]:
df_actual=df_no_duplicates.loc[(df_no_duplicates['year_of_release'] >=2000) & (df_no_duplicates['year_of_release'] <=2013)]

In [31]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 3394 to 9260
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12781 non-null  object 
 1   platform         12781 non-null  object 
 2   year_of_release  12781 non-null  int64  
 3   genre            12781 non-null  object 
 4   na_sales         12781 non-null  float64
 5   eu_sales         12781 non-null  object 
 6   jp_sales         12781 non-null  object 
 7   other_sales      12781 non-null  float64
 8   critic_score     7169 non-null   float64
 9   user_score       6483 non-null   float64
 10  rating           8723 non-null   object 
dtypes: float64(4), int64(1), object(6)
memory usage: 1.2+ MB


In [32]:
df_actual=df_actual.copy()

---

## 4. Категоризация данных
    
- Проводим категоризацию данных по оценкам пользователей


In [33]:
def user_categories(cat):
    if cat <3:
        return 'низкая оценка'
    elif cat <8:
        return 'средняя оценка'
    elif cat <=10:
        return 'высокая оценка'
df_actual['user_category']=df_actual['user_score'].apply(user_categories)

- Проводим категоризацию данных по оценкам критиков.

In [34]:
def critic_categories(cat):
    if cat <30:
        return 'низкая оценка'
    elif cat <80:
        return 'средняя оценка'
    elif cat <=100:
        return 'высокая оценка'
df_actual['critic_category']=df_actual['critic_score'].apply(critic_categories)

In [35]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 3394 to 9260
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12781 non-null  object 
 1   platform         12781 non-null  object 
 2   year_of_release  12781 non-null  int64  
 3   genre            12781 non-null  object 
 4   na_sales         12781 non-null  float64
 5   eu_sales         12781 non-null  object 
 6   jp_sales         12781 non-null  object 
 7   other_sales      12781 non-null  float64
 8   critic_score     7169 non-null   float64
 9   user_score       6483 non-null   float64
 10  rating           8723 non-null   object 
 11  user_category    6483 non-null   object 
 12  critic_category  7169 non-null   object 
dtypes: float64(4), int64(1), object(8)
memory usage: 1.4+ MB


In [36]:
#Кол-вр игры по категориям среди пользователей
grouped_df_user = df_actual.groupby('user_category')['name'].count()
print(grouped_df_user)

user_category
высокая оценка    2286
низкая оценка      116
средняя оценка    4081
Name: name, dtype: int64


In [37]:
#Кол-вр игры по категориям среди критиков
grouped_df_critic = df_actual.groupby('critic_category')['name'].count()
print(grouped_df_critic)

critic_category
высокая оценка    1692
низкая оценка       55
средняя оценка    5422
Name: name, dtype: int64


- Топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [38]:
top_7 = df_actual.groupby('platform').size().sort_values(ascending=False)
print(top_7.head(7))

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
dtype: int64


За период с 2000 по 2013 год в Топ-7 платформ по количеству выпущенных игр вошли такие платформы как: PS2, DS, Wii, PSP, X360, PS3, GBA.

Лидирующие позиции с небольшим отрывом заняли платформы PS2 и DS, заняв большую долю выпущенных игр.                                    

---

## 5. Итоговый вывод


Были загружены данные /datasets/new_games.csv. Они содержат 11 столбцов и 16956 строк, в которых представлена информация об играх. При первичном знакомстве с данными и их предобработкой получили такие результаты:
- В семи столбцах ('Name', 'Genre', 'EU sales', 'JP sales', 'Critic Score', 'User Score', 'Rating') были обнаружены пропущенные значения. Максимальное значение пропущенных данных в столбце 'User Score' - 54%, на втором месте столбец 'Critic Score' - 51%, на третем месте столбец 'Rating' - 40%
- Для оптимизации работы с данными в датафрейме были произведены следующие изменения типов данных:
    - 'Year of Release': тип данных изменён с float64 на int16.
    - 'EU sales' и 'JP sales': тип данных изменён с object на float64.
    - 'User Score': тип данных изменён с object на float64.
    - 'Rating': тип данных изменён с object на category.
- Обработаны пропущенные значения. 
- Проведена нормализация данных с текстовыми значениями с целью исключения неявных дубликатов, а также выявлен 235 явных дубликатов. После удаления явных дубликатов осталось 16679 строк.
- Данные отфильтрованы по периоду с 2000 по 2013 год включительно, после чего осталась 12781 строка.
- Для дополнительной работы с данными были созданы дополнительные столбцы:
    - user_category: игры по категориям оценки пользователей
    - critic_category: игры по категориям оценки критиков
    - А также данные столбцы сгруппированы по количеству игр в каждой категории среди пользователей и критиков
- Выделен топ-7 платформ по количеству игр, выпущенных за весь актуальный период 